In [1]:
pip install lxml

Note: you may need to restart the kernel to use updated packages.


## Installing Geopy and Folium

In [2]:
#installing libraries
!conda install -c conda-forge geopy --yes 
!conda install -c conda-forge folium=0.5.0

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.11
  latest version: 4.8.1

Please update conda by running

    $ conda update -n base -c defaults conda



# All requested packages already installed.

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.11
  latest version: 4.8.1

Please update conda by running

    $ conda update -n base -c defaults conda



# All requested packages already installed.



## Importing other essential libaries 

In [3]:
#Importing essential libraries

import pandas as pd
import numpy as np 
import json
from geopy.geocoders import Nominatim 
import requests 
from pandas.io.json import json_normalize 
import matplotlib.cm as cm
import matplotlib.colors as colors

from sklearn.cluster import KMeans

import folium 

print('Libraries imported.')

Libraries imported.


## Creating a data frame

In [4]:
#Creating a dataframe

dfs = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')
for df in dfs:
    print(df)

    Postcode           Borough          Neighbourhood
0        M1A      Not assigned           Not assigned
1        M2A      Not assigned           Not assigned
2        M3A        North York              Parkwoods
3        M4A        North York       Victoria Village
4        M5A  Downtown Toronto           Harbourfront
..       ...               ...                    ...
282      M8Z         Etobicoke              Mimico NW
283      M8Z         Etobicoke     The Queensway West
284      M8Z         Etobicoke  Royal York South West
285      M8Z         Etobicoke         South of Bloor
286      M9Z      Not assigned           Not assigned

[287 rows x 3 columns]
                                                  0   \
0                                                NaN   
1  NL NS PE NB QC ON MB SK AB BC NU/NT YT A B C E...   
2                                                 NL   
3                                                  A   

                                               

## Cleansing the dataframe

In [5]:
#Creating a dataframe after eliminating unnecessary tables
Canada=dfs[0]
Canada

Postcode           Borough          Neighbourhood
0        M1A      Not assigned           Not assigned
1        M2A      Not assigned           Not assigned
2        M3A        North York              Parkwoods
3        M4A        North York       Victoria Village
4        M5A  Downtown Toronto           Harbourfront
..       ...               ...                    ...
282      M8Z         Etobicoke              Mimico NW
283      M8Z         Etobicoke     The Queensway West
284      M8Z         Etobicoke  Royal York South West
285      M8Z         Etobicoke         South of Bloor
286      M9Z      Not assigned           Not assigned

[287 rows x 3 columns]

In [6]:
#Removing rows which have do not have an assigned Borough

import numpy as np

Canada['Borough']=Canada['Borough'].replace('Not assigned',np.nan)
Canada=Canada.dropna(axis=0,subset=['Borough'])
Canada

Postcode           Borough             Neighbourhood
2        M3A        North York                 Parkwoods
3        M4A        North York          Victoria Village
4        M5A  Downtown Toronto              Harbourfront
5        M6A        North York          Lawrence Heights
6        M6A        North York            Lawrence Manor
..       ...               ...                       ...
281      M8Z         Etobicoke  Kingsway Park South West
282      M8Z         Etobicoke                 Mimico NW
283      M8Z         Etobicoke        The Queensway West
284      M8Z         Etobicoke     Royal York South West
285      M8Z         Etobicoke            South of Bloor

[210 rows x 3 columns]

In [7]:
#Aggregating neigbourhoods who have same postcode
Canada_1=Canada.groupby('Postcode').agg(lambda x:','.join(set(x)))
Canada_1

Borough                                      Neighbourhood
Postcode                                                                
M1B       Scarborough                                      Malvern,Rouge
M1C       Scarborough               Highland Creek,Rouge Hill,Port Union
M1E       Scarborough                    Morningside,Guildwood,West Hill
M1G       Scarborough                                             Woburn
M1H       Scarborough                                          Cedarbrae
...               ...                                                ...
M9N              York                                             Weston
M9P         Etobicoke                                          Westmount
M9R         Etobicoke  Richview Gardens,Kingsview Village,St. Phillip...
M9V         Etobicoke  Mount Olive,South Steeles,Thistletown,Silverst...
M9W         Etobicoke                                          Northwest

[103 rows x 2 columns]

In [8]:
#Filling up not assigned Neighbourhoods with boroughs

Canada_1.loc[Canada_1['Neighbourhood']=="Not assigned",'Neighbourhood']=Canada_1.loc[Canada_1['Neighbourhood']=="Not assigned",'Borough']
Canada_1

Borough                                      Neighbourhood
Postcode                                                                
M1B       Scarborough                                      Malvern,Rouge
M1C       Scarborough               Highland Creek,Rouge Hill,Port Union
M1E       Scarborough                    Morningside,Guildwood,West Hill
M1G       Scarborough                                             Woburn
M1H       Scarborough                                          Cedarbrae
...               ...                                                ...
M9N              York                                             Weston
M9P         Etobicoke                                          Westmount
M9R         Etobicoke  Richview Gardens,Kingsview Village,St. Phillip...
M9V         Etobicoke  Mount Olive,South Steeles,Thistletown,Silverst...
M9W         Etobicoke                                          Northwest

[103 rows x 2 columns]

In [9]:
Canada_1.shape

(103, 2)

## Extracting latitude-longitude data

In [10]:
#Extracting Latitude Longitude data

LatLong=pd.read_csv("http://cocl.us/Geospatial_data")
LatLong

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [11]:
#Setting index to dataframes

LatLong.set_index("Postal Code")


Latitude  Longitude
Postal Code                      
M1B          43.806686 -79.194353
M1C          43.784535 -79.160497
M1E          43.763573 -79.188711
M1G          43.770992 -79.216917
M1H          43.773136 -79.239476
...                ...        ...
M9N          43.706876 -79.518188
M9P          43.696319 -79.532242
M9R          43.688905 -79.554724
M9V          43.739416 -79.588437
M9W          43.706748 -79.594054

[103 rows x 2 columns]

## Merging Neighbourhood data with latitude-longitude data

In [12]:
#Merging dataframes

Canada_co=Canada_1.merge(LatLong,left_on="Postcode",right_on="Postal Code")
Canada_co

Borough                                      Neighbourhood  \
0    Scarborough                                      Malvern,Rouge   
1    Scarborough               Highland Creek,Rouge Hill,Port Union   
2    Scarborough                    Morningside,Guildwood,West Hill   
3    Scarborough                                             Woburn   
4    Scarborough                                          Cedarbrae   
..           ...                                                ...   
98          York                                             Weston   
99     Etobicoke                                          Westmount   
100    Etobicoke  Richview Gardens,Kingsview Village,St. Phillip...   
101    Etobicoke  Mount Olive,South Steeles,Thistletown,Silverst...   
102    Etobicoke                                          Northwest   

    Postal Code   Latitude  Longitude  
0           M1B  43.806686 -79.194353  
1           M1C  43.784535 -79.160497  
2           M1E  43.763573 -79.188711  
3           M1G  43.770992 -79.216917  
4           M1H  43.773136 -79.239476  
..          ...        ...        ...  
98          M9N  43.706876 -79.518188  
99          M9P  43.696319 -79.532242  
100         M9R  43.688905 -79.554724  
101         M9V  43.739416 -79.588437  
102         M9W  43.706748 -79.594054  

[103 rows x 5 columns]

In [13]:
#Set index to postal code

Canada_co.set_index("Postal Code")

Borough                                      Neighbourhood  \
Postal Code                                                                   
M1B          Scarborough                                      Malvern,Rouge   
M1C          Scarborough               Highland Creek,Rouge Hill,Port Union   
M1E          Scarborough                    Morningside,Guildwood,West Hill   
M1G          Scarborough                                             Woburn   
M1H          Scarborough                                          Cedarbrae   
...                  ...                                                ...   
M9N                 York                                             Weston   
M9P            Etobicoke                                          Westmount   
M9R            Etobicoke  Richview Gardens,Kingsview Village,St. Phillip...   
M9V            Etobicoke  Mount Olive,South Steeles,Thistletown,Silverst...   
M9W            Etobicoke                                          Northwest   

              Latitude  Longitude  
Postal Code                        
M1B          43.806686 -79.194353  
M1C          43.784535 -79.160497  
M1E          43.763573 -79.188711  
M1G          43.770992 -79.216917  
M1H          43.773136 -79.239476  
...                ...        ...  
M9N          43.706876 -79.518188  
M9P          43.696319 -79.532242  
M9R          43.688905 -79.554724  
M9V          43.739416 -79.588437  
M9W          43.706748 -79.594054  

[103 rows x 4 columns]

## Locating co-ordinates of Toronto

In [14]:
#Co-ordinates of Toronto
address = 'Toronto'

geolocator = Nominatim(user_agent="foursquare_agent")

location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Totonto is {}, {}.'.format(latitude, longitude))


The geograpical coordinate of Totonto is 43.653963, -79.387207.


## Generating map of Toronto

In [15]:
# Generating map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighbourhood in zip(Canada_co['Latitude'], Canada_co['Longitude'], Canada_co['Borough'], Canada_co['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='red',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Toronto)  
    
map_Toronto

## Generating list of Borough containing Toronto

In [16]:
#Generating list of borough containing keyword Toronto

borough_names = list(Canada_co.Borough.unique())

toronto = []

for x in borough_names:
    if "toronto" in x.lower():
        toronto.append(x)
        
toronto

['East Toronto', 'Central Toronto', 'Downtown Toronto', 'West Toronto']

## Creating dataframe of with borough containing the keyword Toronto

In [17]:
#Creating Dataframes 
toronto_df = Canada_co[Canada_co['Borough'].isin(toronto)].reset_index(drop=True)
print(toronto_df.shape)
toronto_df.head(39)

(39, 5)


Borough                                      Neighbourhood  \
0       East Toronto                                        The Beaches   
1       East Toronto                        Riverdale,The Danforth West   
2       East Toronto                      The Beaches West,India Bazaar   
3       East Toronto                                    Studio District   
4    Central Toronto                                      Lawrence Park   
5    Central Toronto                                   Davisville North   
6    Central Toronto                                 North Toronto West   
7    Central Toronto                                         Davisville   
8    Central Toronto                         Summerhill East,Moore Park   
9    Central Toronto  Summerhill West,Forest Hill SE,Deer Park,South...   
10  Downtown Toronto                                           Rosedale   
11  Downtown Toronto                         St. James Town,Cabbagetown   
12  Downtown Toronto                               Church and Wellesley   
13  Downtown Toronto                                       Harbourfront   
14  Downtown Toronto                            Ryerson,Garden District   
15  Downtown Toronto                                     St. James Town   
16  Downtown Toronto                                        Berczy Park   
17  Downtown Toronto                                 Central Bay Street   
18  Downtown Toronto                             Adelaide,Richmond,King   
19  Downtown Toronto    Harbourfront East,Toronto Islands,Union Station   
20  Downtown Toronto            Design Exchange,Toronto Dominion Centre   
21  Downtown Toronto                      Commerce Court,Victoria Hotel   
22   Central Toronto                                           Roselawn   
23   Central Toronto                 Forest Hill West,Forest Hill North   
24   Central Toronto                  Yorkville,The Annex,North Midtown   
25  Downtown Toronto                      University of Toronto,Harbord   
26  Downtown Toronto            Kensington Market,Grange Park,Chinatown   
27  Downtown Toronto  South Niagara,Bathurst Quay,King and Spadina,R...   
28  Downtown Toronto                    Stn A PO Boxes 25 The Esplanade   
29  Downtown Toronto              Underground city,First Canadian Place   
30  Downtown Toronto                                           Christie   
31      West Toronto                        Dovercourt Village,Dufferin   
32      West Toronto                            Little Portugal,Trinity   
33      West Toronto         Exhibition Place,Parkdale Village,Brockton   
34      West Toronto                       High Park,The Junction South   
35      West Toronto                              Parkdale,Roncesvalles   
36      West Toronto                                  Swansea,Runnymede   
37  Downtown Toronto                                       Queen's Park   
38      East Toronto  Business Reply Mail Processing Centre 969 Eastern   

   Postal Code   Latitude  Longitude  
0          M4E  43.676357 -79.293031  
1          M4K  43.679557 -79.352188  
2          M4L  43.668999 -79.315572  
3          M4M  43.659526 -79.340923  
4          M4N  43.728020 -79.388790  
5          M4P  43.712751 -79.390197  
6          M4R  43.715383 -79.405678  
7          M4S  43.704324 -79.388790  
8          M4T  43.689574 -79.383160  
9          M4V  43.686412 -79.400049  
10         M4W  43.679563 -79.377529  
11         M4X  43.667967 -79.367675  
12         M4Y  43.665860 -79.383160  
13         M5A  43.654260 -79.360636  
14         M5B  43.657162 -79.378937  
15         M5C  43.651494 -79.375418  
16         M5E  43.644771 -79.373306  
17         M5G  43.657952 -79.387383  
18         M5H  43.650571 -79.384568  
19         M5J  43.640816 -79.381752  
20         M5K  43.647177 -79.381576  
21         M5L  43.648198 -79.379817  
22         M5N  43.711695 -79.416936  
23         M5P  43.696948 -79.411307  
24         M5R  43.672710 -79.405678 

## Extracting locational data from Four Square

In [26]:
CLIENT_ID = '3H5Y5LXWNZ0EBEX1HCK4PWBJWR3CN1U23U4VK5GBR0VTFA3C' # your Foursquare ID
CLIENT_SECRET = 'J11HO1VLJLTGOKIWSZUBI1OOCZS1J2EKOF2DG032JNRG2VKL' # your Foursquare Secret
VERSION = '20180604' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 3H5Y5LXWNZ0EBEX1HCK4PWBJWR3CN1U23U4VK5GBR0VTFA3C
CLIENT_SECRET:J11HO1VLJLTGOKIWSZUBI1OOCZS1J2EKOF2DG032JNRG2VKL


In [27]:
def getNearbyVenues(names, latitudes, longitudes, radius=500, LIMIT=100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&r0adius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]["groups"][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [30]:
#Getting venues near Toronto

toronto_venues = getNearbyVenues(names=toronto_df['Neighbourhood'],
                                   latitudes=toronto_df['Latitude'],
                                   longitudes=toronto_df['Longitude'])

The Beaches
Riverdale,The Danforth West
The Beaches West,India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Summerhill East,Moore Park
Summerhill West,Forest Hill SE,Deer Park,South Hill,Rathnelly
Rosedale
St. James Town,Cabbagetown
Church and Wellesley
Harbourfront
Ryerson,Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide,Richmond,King
Harbourfront East,Toronto Islands,Union Station
Design Exchange,Toronto Dominion Centre
Commerce Court,Victoria Hotel
Roselawn
Forest Hill West,Forest Hill North
Yorkville,The Annex,North Midtown
University of Toronto,Harbord
Kensington Market,Grange Park,Chinatown
South Niagara,Bathurst Quay,King and Spadina,Railway Lands,CN Tower,Harbourfront West,Island airport
Stn A PO Boxes 25 The Esplanade
Underground city,First Canadian Place
Christie
Dovercourt Village,Dufferin
Little Portugal,Trinity
Exhibition Place,Parkdale Village,Brockton
High Park,The Junction South
Parkdale,Roncesvalles
Swansea,R

In [31]:
toronto_venues.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0   The Beaches               43.676357               -79.293031   
1   The Beaches               43.676357               -79.293031   
2   The Beaches               43.676357               -79.293031   
3   The Beaches               43.676357               -79.293031   
4   The Beaches               43.676357               -79.293031   

               Venue  Venue Latitude  Venue Longitude       Venue Category  
0    The Fox Theatre       43.672801       -79.287272  Indie Movie Theater  
1     The Beech Tree       43.680493       -79.288846            Gastropub  
2    Ed's Real Scoop       43.672630       -79.287993       Ice Cream Shop  
3     Bagels On Fire       43.672864       -79.286784           Bagel Shop  
4  Beaches Bake Shop       43.680363       -79.289692               Bakery

## Analyzing data extracted from FourSquare

In [32]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Afghan Restaurant  Airport  Airport Lounge  \
0   The Beaches                  0        0               0   
1   The Beaches                  0        0               0   
2   The Beaches                  0        0               0   
3   The Beaches                  0        0               0   
4   The Beaches                  0        0               0   

   American Restaurant  Amphitheater  Animal Shelter  Antique Shop  Aquarium  \
0                    0             0               0             0         0   
1                    0             0               0             0         0   
2                    0             0               0             0         0   
3                    0             0               0             0         0   
4                    0             0               0             0         0   

   Art Gallery  ...  Vegetarian / Vegan Restaurant  Video Store  \
0            0  ...                              0            0   
1            0  ...                              0            0   
2            0  ...                              0            0   
3            0  ...                              0            0   
4            0  ...                              0            0   

   Vietnamese Restaurant  Whisky Bar  Wine Bar  Wine Shop  Wings Joint  \
0                      0           0         0          0            0   
1                      0           0         0          0            0   
2                      0           0         0          0            0   
3                      0           0         0          0            0   
4                      0           0         0          0            0   

   Women's Store  Yoga Studio  Zoo  
0              0            0    0  
1              0            0    0  
2              0            0    0  
3              0            0    0  
4              0            0    0  

[5 rows x 267 columns]

In [33]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Afghan Restaurant  \
0                              Adelaide,Richmond,King               0.00   
1                                         Berczy Park               0.00   
2   Business Reply Mail Processing Centre 969 Eastern               0.00   
3                                  Central Bay Street               0.00   
4                                            Christie               0.00   
5                                Church and Wellesley               0.01   
6                       Commerce Court,Victoria Hotel               0.00   
7                                          Davisville               0.00   
8                                    Davisville North               0.00   
9             Design Exchange,Toronto Dominion Centre               0.00   
10                        Dovercourt Village,Dufferin               0.00   
11         Exhibition Place,Parkdale Village,Brockton               0.00   
12                 Forest Hill West,Forest Hill North               0.00   
13                                       Harbourfront               0.00   
14    Harbourfront East,Toronto Islands,Union Station               0.00   
15                       High Park,The Junction South               0.00   
16            Kensington Market,Grange Park,Chinatown               0.00   
17                                      Lawrence Park               0.00   
18                            Little Portugal,Trinity               0.00   
19                                 North Toronto West               0.00   
20                              Parkdale,Roncesvalles               0.00   
21                                       Queen's Park               0.00   
22                        Riverdale,The Danforth West               0.00   
23                                           Rosedale               0.01   
24                                           Roselawn               0.00   
25                            Ryerson,Garden District               0.00   
26  South Niagara,Bathurst Quay,King and Spadina,R...               0.00   
27                                     St. James Town               0.00   
28                         St. James Town,Cabbagetown               0.00   
29                    Stn A PO Boxes 25 The Esplanade               0.00   
30                                    Studio District               0.00   
31                         Summerhill East,Moore Park               0.00   
32  Summerhill West,Forest Hill SE,Deer Park,South...               0.00   
33                                  Swansea,Runnymede               0.00   
34                                        The Beaches               0.00   
35                      The Beaches West,India Bazaar               0.00   
36              Underground city,First Canadian Place               0.00   
37                      University of Toronto,Harbord               0.00   
38                  Yorkville,The Annex,North Midtown               0.00   

    Airport  Airport Lounge  American Restaurant  Amphitheater  \
0      0.00            0.00                 0.02          0.00   
1      0.00            0.00                 0.00          0.00   
2      0.00            0.00                 0.02          0.00   
3      0.00            0.00                 0.01          0.00   
4      0.00            0.00                 0.01          0.00   
5      0.00            0.00                 0.01          0.00   
6      0.00            0.00                 0.02          0.00   
7      0.00            0.00                 0.00          0.00   
8      0.00            0.00                 0.00          0.00   
9      0.00            0.00                 0.03          0.00   
10     0.00            0.00                 0.00          0.00   
11     0.00            0.00                 0.02          0.00   
12     0.00            0.00                 0.01          0.00   
13     0.00            0.00                 0.00          0.00   
14     0.

In [34]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide,Richmond,King----
             venue  freq
0      Coffee Shop  0.08
1             Café  0.05
2              Bar  0.04
3       Steakhouse  0.04
4  Thai Restaurant  0.03


----Berczy Park----
         venue  freq
0  Coffee Shop  0.10
1         Café  0.05
2     Beer Bar  0.04
3        Hotel  0.04
4   Restaurant  0.04


----Business Reply Mail Processing Centre 969 Eastern----
         venue  freq
0  Coffee Shop  0.07
1      Brewery  0.06
2         Park  0.06
3         Café  0.06
4       Bakery  0.05


----Central Bay Street----
                 venue  freq
0          Coffee Shop  0.13
1   Italian Restaurant  0.04
2      Bubble Tea Shop  0.03
3  Japanese Restaurant  0.03
4                 Café  0.03


----Christie----
               venue  freq
0               Café  0.11
1        Coffee Shop  0.07
2                Bar  0.05
3  Korean Restaurant  0.05
4      Grocery Store  0.04


----Church and Wellesley----
                 venue  freq
0          Coffee Shop  0.11
1  Japanese 

In [35]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [36]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                             Adelaide,Richmond,King           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2  Business Reply Mail Processing Centre 969 Eastern           Coffee Shop   
3                                 Central Bay Street           Coffee Shop   
4                                           Christie                  Café   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café                   Bar            Steakhouse   
1                  Café                 Hotel              Beer Bar   
2               Brewery                  Park                  Café   
3    Italian Restaurant        Ice Cream Shop        Clothing Store   
4           Coffee Shop                   Bar     Korean Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0      Asian Restaurant       Thai Restaurant               Theater   
1            Restaurant          Cocktail Bar        Breakfast Spot   
2                Bakery                 Beach     Indian Restaurant   
3                Bakery                  Café       Bubble Tea Shop   
4    Italian Restaurant         Grocery Store     Indian Restaurant   

  8th Most Common Venue          9th Most Common Venue 10th Most Common Venue  
0      Sushi Restaurant                          Hotel                    Gym  
1   Japanese Restaurant                           Park     Italian Restaurant  
2           Pizza Place                            Bar     Italian Restaurant  
3   Japanese Restaurant                      Gastropub                    Gym  
4          Dessert Shop  Vegetarian / Vegan Restaurant         Ice Cream Shop

## Finding K-means and creating clusters

In [38]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 2, 4, 1, 3, 1, 2, 1, 1, 2], dtype=int32)

In [39]:
# add clustering labels
neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

Borough                  Neighbourhood Postal Code   Latitude  \
0     East Toronto                    The Beaches         M4E  43.676357   
1     East Toronto    Riverdale,The Danforth West         M4K  43.679557   
2     East Toronto  The Beaches West,India Bazaar         M4L  43.668999   
3     East Toronto                Studio District         M4M  43.659526   
4  Central Toronto                  Lawrence Park         M4N  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               4                  Park           Coffee Shop   
1 -79.352188               4      Greek Restaurant                  Café   
2 -79.315572               4                  Park                  Café   
3 -79.340923               4           Coffee Shop                  Café   
4 -79.388790               1           Coffee Shop    Italian Restaurant   

  3rd Most Common Venue  4th Most Common Venue  5th Most Common Venue  \
0                 Beach                   Café                    Pub   
1                  Park  Vietnamese Restaurant                 Bakery   
2           Coffee Shop                  Beach                Brewery   
3                  Park                Brewery  Vietnamese Restaurant   
4                  Park       Sushi Restaurant                   Café   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0        Breakfast Spot        Ice Cream Shop                Bakery   
1                   Pub    Italian Restaurant   American Restaurant   
2    Italian Restaurant                Bakery           Pizza Place   
3                Bakery     French Restaurant                 Diner   
4                Bakery        Ice Cream Shop          Burger Joint   

  9th Most Common Venue 10th Most Common Venue  
0     Indian Restaurant              BBQ Joint  
1        Ice Cream Shop            Coffee Shop  
2     Indian Restaurant    American Restaurant  
3                   Bar        Thai Restaurant  
4           Supermarket            Yoga Studio

In [40]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [41]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels  \
8                          Summerhill East,Moore Park               0   
9   Summerhill West,Forest Hill SE,Deer Park,South...               0   
10                                           Rosedale               0   
27  South Niagara,Bathurst Quay,King and Spadina,R...               0   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
8     Italian Restaurant                  Park                  Café   
9     Italian Restaurant                  Café                  Park   
10                  Park           Coffee Shop    Italian Restaurant   
27                  Park                  Café           Coffee Shop   

   4th Most Common Venue          5th Most Common Venue 6th Most Common Venue  \
8       Sushi Restaurant                         Bakery          Dessert Shop   
9       Sushi Restaurant  Vegetarian / Vegan Restaurant           Coffee Shop   
10                  Café                            Spa     Indian Restaurant   
27                 Hotel             Italian Restaurant                   Gym   

        7th Most Common Venue 8th Most Common Venue  \
8           Indian Restaurant            Restaurant   
9   Middle Eastern Restaurant           Yoga Studio   
10               Gourmet Shop         Grocery Store   
27                    Brewery        Scenic Lookout   

         9th Most Common Venue 10th Most Common Venue  
8                  Coffee Shop          Grocery Store  
9   Modern European Restaurant             Bagel Shop  
10                       Hotel       Sushi Restaurant  
27                    Aquarium       Baseball Stadium

In [42]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels 1st Most Common Venue  \
4                        Lawrence Park               1           Coffee Shop   
5                     Davisville North               1           Coffee Shop   
6                   North Toronto West               1           Coffee Shop   
7                           Davisville               1           Coffee Shop   
12                Church and Wellesley               1           Coffee Shop   
14             Ryerson,Garden District               1           Coffee Shop   
17                  Central Bay Street               1           Coffee Shop   
22                            Roselawn               1           Coffee Shop   
23  Forest Hill West,Forest Hill North               1           Coffee Shop   
37                        Queen's Park               1           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
4     Italian Restaurant                  Park      Sushi Restaurant   
5     Italian Restaurant                  Café                Bakery   
6     Italian Restaurant                  Café                Bakery   
7                 Bakery    Italian Restaurant                  Café   
12   Japanese Restaurant          Burger Joint                  Café   
14        Cosmetics Shop            Restaurant        Clothing Store   
17    Italian Restaurant        Ice Cream Shop        Clothing Store   
22    Italian Restaurant                  Café                Bakery   
23    Italian Restaurant      Sushi Restaurant                  Park   
37        Sandwich Place    Italian Restaurant                  Café   

        5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
4                        Café                Bakery        Ice Cream Shop   
5           Indian Restaurant                  Park                   Gym   
6                        Park      Sushi Restaurant                Garden   
7           Indian Restaurant                  Park                   Gym   
12                 Restaurant      Sushi Restaurant                   Gym   
14  Middle Eastern Restaurant  Fast Food Restaurant              Tea Room   
17                     Bakery                  Café       Bubble Tea Shop   
22        Japanese Restaurant   Sporting Goods Shop             Bookstore   
23                  Gastropub            Restaurant   Japanese Restaurant   
37                       Park             Gastropub                   Gym   

   8th Most Common Venue     9th Most Common Venue     10th Most Common Venue  
4           Burger Joint               Supermarket                Yoga Studio  
5             Restaurant          Asian Restaurant                  Gastropub  
6           Skating Rink       Japanese Restaurant       Fast Food Restaurant  
7           Dessert Shop       Japanese Restaurant                Yoga Studio  
12          Liquor Store  Mediterranean Restaurant                Men's Store  
14   Japanese Restaurant                      Café           Sushi Restaurant  
17   Japanese Restaurant                 Gastropub                        Gym  
22            Bagel Shop         Food & Drink Shop              Deli / Bodega  
23                  Café                    Bakery  Middle Eastern Restaurant  
37         Burrito Place              Burger Joint           Sushi Restaurant

In [43]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels  \
13                                     Harbourfront               2   
15                                   St. James Town               2   
16                                      Berczy Park               2   
18                           Adelaide,Richmond,King               2   
19  Harbourfront East,Toronto Islands,Union Station               2   
20          Design Exchange,Toronto Dominion Centre               2   
21                    Commerce Court,Victoria Hotel               2   
28                  Stn A PO Boxes 25 The Esplanade               2   
29            Underground city,First Canadian Place               2   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
13           Coffee Shop                Bakery            Restaurant   
15           Coffee Shop                  Café                 Hotel   
16           Coffee Shop                  Café                 Hotel   
18           Coffee Shop                  Café                   Bar   
19           Coffee Shop                 Hotel              Aquarium   
20           Coffee Shop                 Hotel                  Café   
21           Coffee Shop                  Café                 Hotel   
28           Coffee Shop                  Café            Restaurant   
29           Coffee Shop                  Café                 Hotel   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
13                  Park                  Café                   Pub   
15            Restaurant                Bakery    Seafood Restaurant   
16              Beer Bar            Restaurant          Cocktail Bar   
18            Steakhouse      Asian Restaurant       Thai Restaurant   
19            Restaurant                  Café               Brewery   
20            Restaurant                   Bar    Seafood Restaurant   
21            Restaurant             Gastropub    Seafood Restaurant   
28   Japanese Restaurant              Beer Bar                 Hotel   
29            Steakhouse             Gastropub   American Restaurant   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
13               Theater        Breakfast Spot        Farmers Market   
15        Cosmetics Shop        Breakfast Spot    Italian Restaurant   
16        Breakfast Spot   Japanese Restaurant                  Park   
18               Theater      Sushi Restaurant                 Hotel   
19    Italian Restaurant        Scenic Lookout                  Park   
20   American Restaurant             Gastropub         Deli / Bodega   
21            Steakhouse                   Gym   Japanese Restaurant   
28    Italian Restaurant                Bakery          Cocktail Bar   
29            Restaurant                   Bar    Seafood Restaurant   

   10th Most Common Venue  
13     Italian Restaurant  
15                Theater  
16     Italian Restaurant  
18                    Gym  
19                  Plaza  
20             Steakhouse  
21               Beer Bar  
28                    Gym  
29                    Gym

In [44]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels  \
11                  St. James Town,Cabbagetown               3   
24           Yorkville,The Annex,North Midtown               3   
25               University of Toronto,Harbord               3   
26     Kensington Market,Grange Park,Chinatown               3   
30                                    Christie               3   
31                 Dovercourt Village,Dufferin               3   
32                     Little Portugal,Trinity               3   
33  Exhibition Place,Parkdale Village,Brockton               3   
34                High Park,The Junction South               3   
35                       Parkdale,Roncesvalles               3   
36                           Swansea,Runnymede               3   

   1st Most Common Venue          2nd Most Common Venue 3rd Most Common Venue  \
11           Coffee Shop                           Café   Japanese Restaurant   
24           Coffee Shop                           Café    Italian Restaurant   
25                  Café                         Bakery           Coffee Shop   
26                  Café  Vegetarian / Vegan Restaurant    Chinese Restaurant   
30                  Café                    Coffee Shop                   Bar   
31                  Café             Italian Restaurant           Coffee Shop   
32                  Café                            Bar            Restaurant   
33                  Café                    Coffee Shop            Restaurant   
34                  Café                    Coffee Shop                   Bar   
35                  Café                    Coffee Shop                Bakery   
36                  Café                            Bar    Italian Restaurant   

            4th Most Common Venue  5th Most Common Venue  \
11                           Park                    Pub   
24                     Restaurant    Japanese Restaurant   
25  Vegetarian / Vegan Restaurant              Bookstore   
26                            Bar  Vietnamese Restaurant   
30              Korean Restaurant     Italian Restaurant   
31                            Bar                 Bakery   
32                         Bakery            Pizza Place   
33                            Bar              Gift Shop   
34                         Bakery     Italian Restaurant   
35                     Restaurant                   Park   
36                    Coffee Shop            Pizza Place   

     6th Most Common Venue          7th Most Common Venue  \
11                   Diner                         Bakery   
24                  Bakery  Vegetarian / Vegan Restaurant   
25                     Bar                            Pub   
26     Dumpling Restaurant             Mexican Restaurant   
30           Grocery Store              Indian Restaurant   
31                    Park                 Breakfast Spot   
32             Coffee Shop                   Cocktail Bar   
33  Furniture / Home Store                         Bakery   
34             Pizza Place                        Brewery   
35                     Bar             Italian Restaurant   
36                    Park                         Bakery   

      8th Most Common Venue          9th Most Common Venue  \
11          Thai Restaurant                     Restaurant   
24            Grocery Store              French Restaurant   
25  Comfort Food Restaurant                Thai Restaurant   
26              Coffee Shop                   Dessert Shop   
30             Dessert Shop  Vegetarian / Vegan Restaurant   
31             Cocktail Bar               Sushi Restaurant   
32       Italian Restaurant  Vegetarian / Vegan Restaurant   
33                  Theater                     Theme Park   
34               Restaurant                        Dog Run   
35         Sushi Restaurant                      Gastropub   
36             Liquor Store               Sushi Restaurant   

   10th Most Common Venue  
11              Gastropub  
24          

## Analyzing clusters

In [45]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels  \
0                                         The Beaches               4   
1                         Riverdale,The Danforth West               4   
2                       The Beaches West,India Bazaar               4   
3                                     Studio District               4   
38  Business Reply Mail Processing Centre 969 Eastern               4   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                   Park           Coffee Shop                 Beach   
1       Greek Restaurant                  Café                  Park   
2                   Park                  Café           Coffee Shop   
3            Coffee Shop                  Café                  Park   
38           Coffee Shop               Brewery                  Park   

    4th Most Common Venue  5th Most Common Venue 6th Most Common Venue  \
0                    Café                    Pub        Breakfast Spot   
1   Vietnamese Restaurant                 Bakery                   Pub   
2                   Beach                Brewery    Italian Restaurant   
3                 Brewery  Vietnamese Restaurant                Bakery   
38                   Café                 Bakery                 Beach   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0         Ice Cream Shop                Bakery     Indian Restaurant   
1     Italian Restaurant   American Restaurant        Ice Cream Shop   
2                 Bakery           Pizza Place     Indian Restaurant   
3      French Restaurant                 Diner                   Bar   
38     Indian Restaurant           Pizza Place                   Bar   

   10th Most Common Venue  
0               BBQ Joint  
1             Coffee Shop  
2     American Restaurant  
3         Thai Restaurant  
38     Italian Restaurant

## Observations

The cluster of "The Beaches" and "The Beaches West,India Bazaar" has the highest number of Parks. The clusters "Studio District" and "Business Reply Mail Processing Centre 969 Eastern" has the highest number of coffee shops. Cafe has been consistently found has the 2nd most common place in 3 out of 5 clusters